src/Symfony/Component/Debug/DebugClassLoader.php

```diff
@@ -163,11 +164,21 @@ public function loadClass($class)
                 throw new \RuntimeException(sprintf('Case mismatch between loaded and declared class names: %s vs %s', $class, $name));
             }
 
+            if (preg_match('#\n \* @final( .*)?\r?\n \*(?: @|/$)#s', $refl->getDocComment())) {
+                self::$final[$name] = true;
+            }
+
+            $parent = get_parent_class($class);
+            if ($parent && isset(self::$final[$parent])) {
+                @trigger_error(sprintf('Extending the class %s in %s is deprecated and may break in the next major release.', $parent, $name), E_USER_DEPRECATED);
```

Comment:
if you throw, it means that adding ``@final`` somewhere is a BC break, as it forces everyone to update immediately (breaking the dev environment is as bad as breaking the prod one, as your BC would then be valid only for people not using the dev environment to continue working on their project, which means nobody sane).

And we will never throw an exception here (it would suffer from the same issue for new ``@final`` annotation).
the ``may`` is here because the class being wanna-be-final may be turned into a real final class in the next major version of its package (which has nothing to do with the next Symfony major version) if the goal was to move to real final classes. But some package owners may decide to keep them wanna-be-final forever to allow mocking them, so we cannot be sure that the code will break.